geps/gep-1924.md

```diff
@@ -0,0 +1,76 @@
+
+## The Frontend Role
+
+As described in [GEP-1324](https://gateway-api.sigs.k8s.io/geps/gep-1324), the GAMMA spec classifies the Kubernetes Service resource as having two roles: frontend and backend. From the GEP glossary:
+
+> The "service frontend" refers to how we call a service. In a Service, this is an automatically allocated DNS name (name.namespace.svc.cluster.local) and IP address (ClusterIP). Not all services have frontends - for example, "headless" Services.
+> The "service backend" refers to the target of a service. In Service, this is Endpoints (or EndpointSlice). 
+
+Because the Service resource’s roles are split in this way, it is both a valid parentRef and backendRef for a given xRoute resource. However, Service is not the only resource that can fulfill the frontend role. While the Gateway API spec couldn’t possibly enumerate every existing (and future) frontend-like resource, it can specify a subset of resources that implementations MUST support as parentRefs under as a part of core conformance. Meshes MAY support other implementation-specific resources as parentRefs. The spec maintainers also reserve the right to add additional resources to core conformance as the spec evolves.
+
+## Core Conformance
+
+At present (v0.7.0+), there is only 1 parentRef resources that is under core conformance: Service. The semantics of the Service binding can be found in GEP 1426. 
+
+## Extended Conformance
+
+In addition to Service, there are other optional parentRef resources that, if used by implementations, MUST adhere to the spec’s prescriptions. At the time of writing (v0.70), there are two resources in extended conformance: ServiceImport (part of the mcs-api, currently in alpha) and Gateway . The semantics of ServiceImport parentRef binding can be found in GEP 1748 (Note: Headless ServiceImports are out of scope and not currently a part of the spec). The semantics for Gateways in the context of GAMMA implementations are laid out below:
+
+### `Gateways`
+
+There has been much discussion around cluster local Gateways (i.e. Gateways not associated with a traditional load balancer). While there are various potential UX impairments (e.g. what’s the difference between a GAMMA HTTPRoute with a Gateway parentRef and an ingress implementation’s HTTPRoute?), there is no technical reason why a Gateway cannot be a valid GAMMA parentRef. In this case, the mesh implementation MUST use the ClusterIP of the Gateway for traffic matching (but not necessarily to the exclusion of the External IP).
```

Comment:
This started out being concise and then turned to me rambling:

There is something about this that doesn't quite sit right with me around "MUST use the ClusterIP of the Gateway for traffic matching".

A Gateway doesn't have a ClusterIP, that is a Service construct. It has an Address, which may be local to the cluster (and it may even literally be a ClusterIP, copied from a Service into the Gateway status).

The spec declares that when you connect to the Gateway's address, the configuration on the GW (listeners + routing rules + policy + etc) are actuated.

If we happen to have a GatewayClass in the future that generates a "cluster local IP" (and maybe even a DNS name!) and puts them in the gateway status, we would of course use them and should apply config to them.

---

Really nothing in GAMMA is actually mesh specific. The Gateway API just says a parent is something we are attaching a route to. A Gateway is something with an address we can call. A Service is also something with an address we can call, so of course we can bind to it.

If we were to have a "ClusterIP GatewayClass", a "GAMMA implementation" is really just the same as "A controller than implements the 'ClusterIP GatewayClass'". That happens to be something only really feasible by products that are "service meshes", but you can explain all of Gateway - including 'GAMMA' - without the word "mesh" IMO.

Reply:
> Really nothing in GAMMA is actually mesh specific.
I think that is the struggle.  Somehow it has to be possible to say, I have a Mesh-y thing servicing this or I have an ingress-y thing servicing that.  Without precluding that there can't be a future thing that can do all the things and how do we not make that feel unnatural.

To me the Service remains the logical thing.  My ingress-y thing needs to know that in this case it should send to the Service cluster IP for a mesh-y thing to then take over.  And for this other service there is no mesh-y thing and traffic can be steered direct to the Service pods and the ingress-y thing simply does more work.
I don't think I have an answer, but feel a need to recognize the historic assumptions that I think we are stuck on.

Reply:
> A Gateway doesn't have a ClusterIP, that is a Service construct. It has an Address, which may be local to the cluster (and it may even literally be a ClusterIP, copied from a Service into the Gateway status).

This is a fantastic call-out and an oversight on my part. I agree 100% that Service constructs shouldn't scope creep in the Gateway spec. 

> Somehow it has to be possible to say, I have a Mesh-y thing servicing this or I have an ingress-y thing servicing that.

Would the controllerName in GatewayClass suffice?

Reply:
I don't know what 'mesh-y' is ( and I've been worked on this for 5 years). 

There is a DNS name that resolves to an IP, and requests happen to go where HttpRoute decides. Nothing 'mesh-y' or 'ingress-y' about it - that's how internet has worked long before Istio/meshes and K8S. 

Some Gateway implementation already support 'internal gateway' model ( i.e. IP visible only in the VPC). Service itself can be of LoadBalancer type - also external or internal. Will a HttpRoute attached to a Service with type LoadBalancer/external be any different from a gateway ? How and why ? 

Will a Gateway of type 'internal' be different from a mesh-y ? 

I think the main difference between a Service/LoadBalancer and a Gateway is that Gateway allow additional 
L7 settings and controls. But there is nothing in Gateway that requires 'ingress' ( i.e. internet traffic going into the VPC).

Reply:
I am not sure I grok how a ClusterIP Gateway and Service are different. Are they not just syntax differences for the exact same thing?

Given its not right now a "vague idea" and not an implementation, maybe we just have different ideas of what it is/will be, though.

Reply:
They're not different in the functions they provide, I'm arguing that they're different in the implied usage, because a Gateway is intended to be a thing that sits at the logical edge of a cluster and translates. It's possible (and will almost certainly be the case for Cilium) that the act of passing through a Gateway may end up with your traffic being classified differently (security policy wise) that if it passes through a GAMMA Service binding.

We evidently haven't done a good enough job of communicating that this is the main differentiator of a _Gateway_ to other types of networking construct - the _translation_ between "inside" and "outside". The fact that "inside" and "outside" are murky when talking about ClusterIP Gateways doesn't change the fact that this is the intended behavior.

We've also had this come up with Cilium when people go from a pod _inside_ the cluster to the _outside_ of an Ingress entry - because of the Ingress contract, we're defining that to mean that the original source inside the cluster is stripped. Instead the traffic will pick up the "ingress" identity. If this behavior is undesirable, then we need to ensure that there's another construct available - and that's the point of the Service binding - it doesn't have the implied translation that a Gateway does.

Reply:
Thanks for the discussion! For now, I removed mention of ClusterIP from Gateway discussion